### PR TITLE
snap-bootstrap: add CVM mode

### PR DESCRIPTION
**Pull request by xnox**
Add CVM mode

### CODE REVIEW

File: boot/cmdline.go
Comment (anonymouse64) [inline boot/cmdline.go:64]: we eventually want no mode specified to actually be an error in the initramfs for UC20, is it possible to instead make this `snapd_recovery_mode=run-cloudimg-rootfs-draft-1` specified on the kernel command line ?

**Review by anonymouse64 — COMMENTED** [review]
some small comments

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (xnox) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1380]: no, it has no partition label. it is the FS label, which the code fallsback to. I don't think the cloud images currently set partition label at all for the ESP. Quite often it is UEFI or ESP.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (xnox) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1411]: At the moment, since I am still prototyping I'm trying to make things work as optionally encrypted. I think we will remove optionality once this goes more production ready.

File: boot/cmdline.go
Comment (xnox) [reply boot/cmdline.go:64]: At the moment, I'm struggling to make that work. I am hoping to be able to get there though.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (anonymouse64) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1380]: ah you're right I had forgotten even though I apparently wrote that code to fallback on FS label...

but I still wanted to call this out to confirm as using UEFI filesystem label is different from just "ubuntu-seed" for UC20, but it sounds like that is expected/by-design

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:1393]: this mst.UnverifiedBootModel helper will read the model from ubuntu-boot which this code is not mount even afaict, also I suppose there won't be separate boot dir here?

File: boot/cmdline.go
Comment (pedronis) [inline boot/cmdline.go:46]: what's the plan for this PR, we can't really land these changes as they here in snapd master

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (pedronis) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1393]: also that code both in snapd and secboot ATM won't measure the fact this model is classic but very likely should

**Review by pedronis — COMMENTED** [review]
comments/questions

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (xnox) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1393]: measuring `classic: true` vs `classic: false` would be nice. But i guess change of the measurement which is not backwards compatible? I guess it should measure if classic is true.

Yeah,  I don't know how to specify that all my partition layouts are weird. Cause model in my case should be read from ESP (aka ubuntu-seed in uc20 terms, but UEFI in cloudimg-rootfs terms)

File: boot/cmdline.go
Comment (xnox) [reply boot/cmdline.go:46]: the plan is that this PR will be built into a forked core-initrd, to be used in forked cloud images for now. It is intentionally worded as not mergable into snapd master at the moment.

Comment (pedronis) [reply boot/cmdline.go:46]: thank for clarifying this

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (pedronis) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1393]: we can work something in secboot here and secboot the package about the measuring.

About the partitions and layout, that's more annoying snapd codebase assumes some things are fixed and uses global values for them.

File: boot/cmdline.go
Comment (xnox) [reply boot/cmdline.go:64]: Would it be possible to also allow it as a cmdline option to snap-bootstrap? As in `--default-mode="blah"` then classic kernelefi initrd will call it with `--default-mode=cloudimg-rootfs`, whereas uc20 will call snap-bootstrap with `--default-mode=install`. I guess we might have to do that, if it proves impossible to add cmdline args reliably in classic case.

Comment (pedronis) [reply boot/cmdline.go:64]: yes, that sounds a reasonable approach if needed and if the initrd for now need to diverge anyway

Comment (anonymouse64) [reply boot/cmdline.go:64]: an option like that seems reasonable :+1:

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (xnox) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1404]: I can bind mount ESP as ubuntu-seed and ubuntu-boot if that makes things simpler. Also I somehow thought that model is read from either boot or seed....  How do we measure model in install mode?

Comment (xnox) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1404]: oh, i see what is happening now.

I'm now wondering if i should measure the model from ESP, or if I should include a model in the classic initrd and measure that.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (xnox) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go]: I have now changed the code to be strict and require by-uuid mounting.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (xnox) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go]: The images can be encrypted or not. and we want to support both.

File: boot/cmdline.go
Comment (xnox) [reply boot/cmdline.go]: we have worked out a way to have a built-in kernel cmdline in the kernel.efi. if this will be needed in the future a new pr will be opened.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (xnox) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go]: given that we use archive signing key, not uc20 one. There is no concern about not shipping or measuring models for classic right now. Especially since those are not defined yet (aka classic:true grade:secured)

File: boot/cmdline.go
Comment (xnox) [reply boot/cmdline.go]: PR has been updated, and imho it is ready for review and merging now.

File: boot/initramfs20dirs.go
Comment (mardy) [inline boot/initramfs20dirs.go:84]: nit pick: maybe "where pivot-root will be performed to"?

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (mardy) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:1413]: Just a doubt, please ignore me if this doesn't apply: when we return because of an error, should we unwind the operations successfully performed so far? Is there a possibility that this method fails, then gets executed again (after fixing the cause that made it fail), but still fails because of the previous operations taht we didn't clean up?

Comment (xnox) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1413]: Note that none of the other modes unwind things. By default we reboot if we fail here. Emergency shell & debug shells are not available (requires "dangerous" on the kernel cmdline). Thus there is actually no way to "fixing the cause that made it fail" without a reboot.

Comment (mardy) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1413]: Makes sense, thanks for the explanation!

File: boot/flags.go
Comment (anonymouse64) [inline boot/flags.go:146]:
```suggestion
		// no boot flags are consumed / used on cloudimg mode, so return nothing
```

**Comment by xnox** [comment]
@anonymouse64 when is thing going to be merged?

**Comment by anonymouse64** [comment]
@xnox well for start this now has conflicts, if you fix the conflicts I will re-review this PR and the other one tomorrow

**Comment by xnox** [comment]
@chrisccoulson @anonymouse64 @mvo5  i'm not too sure how to reconcile this. At the time this merge proposal was made, snapd's secboot was out of date, hence i pushed it to a newer revision.

Since then, instead of integrating secboot master, a fork was created called snapshot-4c814e1 which does not contain the code that we need here. And it looks like in secboot master & snapshot-4c814e1 are not merged either.

Do I need to propose secboot changes that are needed for this feature into secboot's snapshot-4c814e1?
Or should snapshot-4c814e1 be merged into secboot master?

**Comment by xnox** [comment]
Everything in secboot master is implemented on par with snapshot-4c814e1, i did check it line by line. So this PR "simply" needs to bump to latest secboot and friends.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (anonymouse64) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:1424]: is it really a supported use case to allow entering the recovery key for a cloud image? How would that work? Is it a common thing for clouds to have interactive consoles? I ask because I don't think GCE does or at least we have had that problem with GCE before

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (anonymouse64) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:1441]: this comment needs to be changed

File: secboot/secboot_tpm.go
Comment (anonymouse64) [inline secboot/secboot_tpm.go:389]: I think this function name needs to be more specific, since it is only used to "prepare" the TPM in one specific mode of operation and it uses the srk.tmpl file which AFAICT we don't use in any other mode

File: secboot/secboot_tpm.go
Comment (anonymouse64) [inline secboot/secboot_tpm.go:403]: I'm not sure what this and the code below is meant to do/ how the overall system is designed, so I will leave out this part from my review, but someone else should review this part for correctness, I don't see any logic errors like not handling an error anything but that isn't good enough :smile:

File: vendor/vendor.json
Comment (anonymouse64) [inline vendor/vendor.json:9]: it would be nice if we could move to new secboot, etc. in a separate PR and then just use the new stuff in this PR, it would be easier to review and I would feel more comfortable that we aren't breaking something in this PR with the new secboot changes

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (xnox) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1424]: Most clouds do, and it is explicit requirement.

Usually one wouldn't do that via serial console, but by detaching disk and moving it to a different VM with a different TPM, meaning one will ssh-in normally but will be unable to unlock the disk as the TPM is different.

File: secboot/secboot_tpm.go
Comment (xnox) [reply secboot/secboot_tpm.go:389]: But it is a generic operation, as defined in the TPM v2 spec. Storage Primary Key (SRK) Templates. It is a generic thing to import anything that has been externally pre-sealed against a given TPM. Nothing about this is specific to a given mode, it is a generic way to import externally pre-sealed secrets for this TPM. Any sort of license / decryption keys can be done this way.

File: secboot/secboot_tpm.go
Comment (xnox) [reply secboot/secboot_tpm.go:403]: Take a blob. Offer the blob to the TPM. Cross fingers and hope that TPM will nom the blob. And thus will allow to unseal secrets that have been presealed for this TPM using this template.

File: vendor/vendor.json
Comment (xnox) [reply vendor/vendor.json:9]: yeah, i'm not sure who is working on revving secboot and friends in snapd master. I guess I can attempt to start a fresh PR that does just that.

Comment (anonymouse64) [reply vendor/vendor.json:9]: Well I can tell you I'm not working on it, but @mvo5 is the best person to ask/coordinate with

**Comment by xnox** [comment]
Tests look good; and don't seem to have issues in things that are changed here.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (anonymouse64) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:147]: do we expect to have different labels eventually? if not, I would rather this not be a parameter but rather just a constant defined somewhere here with a comment that "this is the expected value" etc

**Comment by xnox** [comment]
@pedronis this pull request doesn't depend on any other features or code.... it adds a new mode and uses it. I will rebase this, but I need a timeline when to do it, such that it is reviewed in a timely manner and landed without me continiously rebasing it and it getting nor reviewed nor merged for months.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:1413]: this is not really generic mode classic support, as it's not measuring the model and is calling the prepare function that atm is very specific. So this should probably be renamed to something more specific.

label is very generic term, as we use it also for seed system. Is that a diskLabel?

File: secboot/secboot_tpm.go
Comment (pedronis) [inline secboot/secboot_tpm.go:404]: for consistency this needs to be called TPMPrepare, it's a bit of a vogue name though to me, so a more precise name would be better.

Also this bit need a +1 from Chris I think.

This needs a doc comment as well. It would be good if dir had also a more clear name, what kind of dir.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:1441]: this TODO might not be relevant if this function doesn't try to stay generic for the time being. At this point is probably premature for that anyway.

File: boot/cmdline.go
Comment (pedronis) [inline boot/cmdline.go:45]: there's a conflict in this file atm

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:1413]: this is not unit tested currently

**Review by pedronis — COMMENTED** [review]
I did another pass, it needs a bit more work before being ready for master

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (xnox) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1413]: From initramfs-tools manpage:
 > If a label is used,  as in root=LABEL=rootPart the initrd will search all available devices for a filesystem with the appropriate label, and mount that device as the root filesystem.

From mke2fs manpage
> -L new-volume-label Set the volume label for the filesystem to new-volume-label.  The maximum length of the volume label is 16 bytes.

It's not quite `diskLabel`, nor `gptLabel`, but it is rather specifically `fsLabel`, or maybe even `rootFsLabel`.

Comment (xnox) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1413]: > this is not really generic mode classic support

Given all the other architectural plans, i do agree that this is no longer generic hybrid classic mode. But rather something specific to basic, optionally encrypted, classic cloud image.

What should I call it? `generateMountsModeCloudCVM()` would that be suitable?

In that case I can also drop label argument and make it a constant, that is specific to ModeCloudCVM.

File: secboot/secboot_tpm.go
Comment (xnox) [reply secboot/secboot_tpm.go:404]: > Strike a pose, Vogue!

I guess you mean vague =) but the song reference made my day.

But yes, it is vague, and is opaque as to what sort of things we are attempting to import into tpm, and if they match a given tpm, and it is unknown if they already are known, etc. And yes, this bit of code came from @chrisccoulson already anyway. I'll try to give it more descriptive name and docstring to explain what it does.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (xnox) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1441]: I'll drop it then.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (xnox) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1413]: ack.

File: boot/cmdline.go
Comment (xnox) [reply boot/cmdline.go:45]: ack.

File: secboot/secboot_tpm.go
Comment (pedronis) [reply secboot/secboot_tpm.go:404]: yes, I meant vague not https://www.youtube.com/watch?v=GuJQSAiODqI :)

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (pedronis) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:1413]: yes, something like generateMountsModeRunCloudCVM seems appropriate

**Comment by xnox** [comment]
Although I have merged the old branch history, the commits in use are effectively rewrites/forward ports to more recent codebase, addressing renaming of the functions as per feedback and matching closer newer version of secboot.

I have tried to split out individual pull requests of standalone features used by this pull request i.e. 

https://github.com/snapcore/snapd/pull/11902

and

https://github.com/snapcore/snapd/pull/11989

Not sure how useful that is.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:1534]: mmh, this case is not tested but I see that is true for similar code also in preexisting cases

File: cmd/snap-bootstrap/initramfs_systemd_mount.go
Comment (pedronis) [inline cmd/snap-bootstrap/initramfs_systemd_mount.go:210]: nitpick: I think this would slightly more readable as:

`mounted == !opts.Unmount`

**Review by pedronis — COMMENTED** [review]
looks good, some minor comments, did @chrisccoulson review this at some point?

**Review by alfonsosanchezbeato — APPROVED** [review]
LGTM, I have a couple of suggestions if you want to address those

**Comment by xnox** [comment]
> looks good, some minor comments, did @chrisccoulson review this at some point?

@chrisccoulson wrote the original TpmPrepare function which is now named ProvisionTPMCVM, which i merged into xnox/snapd repo together with my mount changes. See https://github.com/xnox/snapd/pull/2 At this point I only rewrote it to adapt to newer way of secboot function naming scheme.

File: cmd/snap-bootstrap/initramfs_systemd_mount.go
Comment (pedronis) [reply cmd/snap-bootstrap/initramfs_systemd_mount.go:103]: I think supporting only unmounting by where is fine, because of the ambiguity issue mentioned.

File: cmd/snap-bootstrap/cmd_initramfs_mounts_test.go
Comment (xnox) [reply cmd/snap-bootstrap/cmd_initramfs_mounts_test.go:2375]: added an assert that the variable is false prior to flipping it to true, to establish one way ratchet.

However, it is also harmless to call this function multiple times, since subsequent calls become no-op if provisioning was successful.

**Review by pedronis — APPROVED** [review]
thank you, lgtm

**Pull request merged**
